File: iOS/SkyService/Data/Models/WorkspaceId.swift
import DittoSwift

struct WorkspaceId: CustomStringConvertible, ExpressibleByStringLiteral {

    static var dateFormatter: DateFormatter {
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "MM-dd-yyyy"
        return dateFormatter
    }
    private static let delimiter: String = "::"

    var description: String {
        return "\(Self.dateFormatter.string(from: self.departureDate))\(Self.delimiter)\(self.flightNumber)"
    }

    var departureDate: Date
    var flightNumber: String

    var id: String

    init(departureDate: Date, flightNumber: String) {
        self.id = "\(Self.dateFormatter.string(from: departureDate))::\(flightNumber)"
        self.departureDate = departureDate
        self.flightNumber = flightNumber
    }

    init(stringLiteral value: StringLiteralType) {
        self.id = value
        let parts = value.components(separatedBy: Self.delimiter)
        self.departureDate = Self.dateFormatter.date(from: parts[0])!
        self.flightNumber = parts[1]
    }
}

File: Android/app/src/main/java/live/dittolive/skyservice/models/WorkspaceId.kt
package live.dittolive.skyservice.models

import org.joda.time.format.DateTimeFormat
import java.time.LocalDate
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.*

class WorkspaceId {

    var id: String
    var departureDate: Date? = null
    var flightNumber: String? = null

    val description: String?
        get() {
            if (departureDate != null) {
                return "${dateFormatter.print(departureDate!!.time)}::${flightNumber}"
            }

            return null
        }

    companion object {
        private val delimiter: String = "::"
        private val dateFormatter = DateTimeFormat.forPattern("MM-dd-yyyy").withLocale(Locale.ENGLISH);
    }

    constructor(value: String) {
        id = value
        val parts = value.split(delimiter)
        departureDate = dateFormatter.parseDateTime(parts[0]).toDate()
        flightNumber = parts[1]
    }
    constructor(flightNumber: String, departureDate: Date) {
        id = "${dateFormatter.print(departureDate.time)}${delimiter}${flightNumber}"
        this.departureDate = departureDate
        this.flightNumber = flightNumber
    }
}
